### DISCUSSION TITLE
How to check if parallel is actually running in parallel

### DISCUSSION THREAD

**Discussion by MLobatoR** [beehive-lab/TornadoVM#290] [Q&A] [answered]
Hi,

Is there any log that I can activate to make sure that parallel code is actually running in parallel and that is running on GPU?

Thank you in advance for your help,
ML

**Comment by mikepapadim** [comment]
Hello @MLobatoR, you can use the `--threadInfo` flag. Then you should expect to get an output similar to: 

```
Task info: foo.bar
	Backend           : OPENCL
	Device            : NVIDIA GeForce RTX 3070 CL_DEVICE_TYPE_GPU (available)
	Dims              : 1
	Global work offset: [0]
	Global work size  : [8096]
	Local  work size  : [1012, 1, 1]
	Number of workgroups  : [8]

```

with the above debug output you can verify that you code is running in parallel on your available device.


This is an example usage of the above-mentioned flag:
```
tornado --threadInfo  -m tornado.examples/uk.ac.manchester.tornado.examples.vectors.DFTVector
```

**Reply by MLobatoR** [reply]
Hi @mikepapadim,

Is there a way to setup threadInfo on TaskGraph or something like that instead of running via cmd?

```
TaskGraph taskGraph = new TaskGraph("compute") //
                .transferToDevice(DataTransferMode.FIRST_EXECUTION, inReal, inImag) //
                .task("withVectors16", DFTVector::computeDFTVector16, inReal, inImag, outReal, outImag) //
                .transferToHost(DataTransferMode.EVERY_EXECUTION, outReal, outImag);
```

Thank you for your help and fast answer :)

**Reply by jjfumero** [reply, marked as answer]
Hi @MLobatoR , 
   to be more precise, you need to look at the `Global work size  : ` field of the `--theadInfo` option. 
This value indicates the number of threads that were launched on the device. If > 1, then it ran in parallel.

Additionally, the `--threadInfo` option prints the device (e.g., which GPU) was used. 

Regarding your last question, there is a method you can invoke from the execution plan to enable profiling information. This will print the device and some metrics, but not the thread information. 

```java
 // Build executionPlan
TornadoExecutionPlan executionPlan = new TornadoExecutionPlan(immutableTaskGraph);

// enable profiler
executionPlan.withProfiler(ProfilerMode.CONSOLE);

// Execute the plan (default TornadoVM optimization choices)
TornadoExecutionResult executionResult = executionPlan.execute();
```

I think it is a good idea to enable the `threadInfo` from the execution plan. We will discuss this internally and provide a new patch.

**Reply by MLobatoR** [reply]
Hi @jjfumero,

Thank you very much for your help.

**Reply by MLobatoR** [reply]
@jjfumero,

Just one more question. De time displayed is in nano seconds? Or other time unit? Can you clarify please?

```
 "TOTAL_BYTE_CODE_GENERATION": "15066100",
        "TOTAL_DISPATCH_DATA_TRANSFERS_TIME": "2170784",
        "TOTAL_GRAAL_COMPILE_TIME": "245557300",
        "TOTAL_DISPATCH_KERNEL_TIME": "147808",
        "COPY_IN_TIME": "13760",
        "TOTAL_TASK_GRAPH_TIME": "548487800",
        "TOTAL_DRIVER_COMPILE_TIME": "151213400",
        "TOTAL_CODE_GENERATION_TIME": "20175400",
        "TOTAL_KERNEL_TIME": "30720",
        "COPY_OUT_TIME": "3616",
        "TOTAL_COPY_OUT_SIZE_BYTES": "1808",
        "TOTAL_COPY_IN_SIZE_BYTES": "44544",
```

Thank you in advance for your help,
ML

**Reply by jjfumero** [reply]
Yes, the timers are displayed in nanoseconds. 
More info: https://tornadovm.readthedocs.io/en/latest/profiler.html#enable-the-profiler-from-the-command-line
